modules/core/include/opencv2/core/cuda.hpp

```diff
@@ -340,6 +340,179 @@ class CV_EXPORTS_W GpuMat
     Allocator* allocator;
 };
 
+class CV_EXPORTS_W GpuMatND
+{
+public:
+    struct CV_EXPORTS_W DevicePtr
+    {
+        explicit DevicePtr(size_t _size);
+         ~DevicePtr();
+
+        DevicePtr(const DevicePtr&) = delete;
+        DevicePtr& operator=(const DevicePtr&) = delete;
+
+        DevicePtr(DevicePtr&&) = delete;
+        DevicePtr& operator=(DevicePtr&&) = delete;
+
+        uchar* data;
```

Comment:
> DevicePtr

It make sense to move this class on upper level, make it more generic, and change its name:
- `GpuData` (like `UMatData`)
- or `GpuBuffer`
- or `GpuDataContainer`

In the future, we can reuse/share this "container" with existed `GpuMat`.

---

Please add `size_t size` field (size of allocated buffer in bytes) to perform accurate validation checks.

Reply:
`GpuData` is a global struct now with a new member `size_t size`.